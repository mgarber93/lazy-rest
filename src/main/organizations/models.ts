import {AuthoredContent} from '../../models/content'
import {OpenAPI} from 'openapi-types'

export interface ApiCallPlan {
  userGoal: AuthoredContent
  state: object
  steps: PlanStep[]
  step: number
  endpoints: string
  oasSpec: OpenAPI.Document[]
}


export type THttp = 'GET' | 'POST' | 'PATCH' | 'PUT' | 'DELETE'

export interface HttpRequestPlan {
  baseUrl: string;
  path: string;
  method: THttp;
  body?: object | Array<object>;
  headers?: Record<string, string>;
}

export interface CallPlan {
  plannedCalls: HttpRequestPlan[]
  summary: string
}

export interface PlanStep {
  apiId: string                   // used for auth
  background: string
  actionDraft: string
  action?: HttpRequestPlan        // step 1 make from background, draft
  result?: object                 // step 2 make call, parse subset of response (stretch goal)
  resultInterpretation?: string   // step 3 decide to retry, continue to next step, or achieved user goal
}