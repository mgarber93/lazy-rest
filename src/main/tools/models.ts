export interface ToolParameter {
  type: "object" | "string" | "number"
  enum: string[]
  description: string
}

export interface ObjectParameter extends ToolParameter {
  type: "object"
  required: string[]
}

export interface Tool {
  name: string
  description: string
  parameters: Record<string, ToolParameter>
}
