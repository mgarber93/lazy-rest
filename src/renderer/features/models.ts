import {createAsyncThunk, createSlice, PayloadAction} from '@reduxjs/toolkit'
import {TModel} from '../../models/responder'
import {ClientOptions} from 'openai'

const name = 'models'

const initialState = {
  models: [] as TModel[],
  ollamaModels: [] as string[],
  bedrockModels: [] as string[],
  providers: {
    openAi: null as ClientOptions | null,
    ollama: null as object | null,
    bedrock: null as {
      region: string;
      credentials: {
        accessKeyId: string;
        secretAccessKey: string;
      };
    } | null,
  },
  organizations: [""],
}


export const listOpenAiModels = createAsyncThunk(
  `${name}/listModels`,
  async () => {
    const models = await window.main.getModels('openai')
    return models
  },
)

export const listOllamaModels = createAsyncThunk(
  `${name}/listOllamaModels`,
  async () => {
    const models = await window.main.getModels('ollama')
    return models
  },
)

export const listBedrockModels = createAsyncThunk(
  `${name}/listBedrockModels`,
  async () => {
    const models = await window.main.getModels('bedrock')
    return models
  },
)

export const modelsSlice = createSlice({
  name,
  initialState,
  reducers: {
    configureOpenAi(state, action: PayloadAction<ClientOptions>) {
      state.providers.openAi = action.payload
    },
    configureBedrock(state, action: PayloadAction<{
      region: string;
      credentials: {
        accessKeyId: string;
        secretAccessKey: string;
      };
    }>) {
      state.providers.bedrock = action.payload
    },
  },
  extraReducers: (builder) => {
    builder.addCase(listOpenAiModels.fulfilled, (state, action) => {
      state.models = action.payload as TModel[]
    })
    builder.addCase(listOllamaModels.fulfilled, (state, action) => {
      state.ollamaModels = action.payload as TModel[]
    })
    builder.addCase(listBedrockModels.fulfilled, (state, action) => {
      state.bedrockModels = action.payload as string[]
    })
  },
})


export const {configureOpenAi, configureBedrock} = modelsSlice.actions
