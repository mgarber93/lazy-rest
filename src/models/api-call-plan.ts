import {v4} from 'uuid'

export interface ApiCallPlan {
  steps: SequenceActivity[]
}

export type THttp = 'GET' | 'POST' | 'PATCH' | 'PUT' | 'DELETE'

export interface HttpRequestPlan {
  name?: string;
  httpVerb: THttp;
  url: string;
  queryParams?: object;
  headers: Record<string, string>;
  body?: object | string
  response?: HttpResponse
}

export interface HttpResponse<T = any> {
  data: T
  status: number
  interpretation?: string
}

export enum ProgressStage {
  active = 'active',
  draft = 'draft',
  planable = 'planable',
  done = 'done',
}

/**
 * An action in a sequence that allows for planning
 */
export interface SequenceActivity {
  id: string
  progressStage: ProgressStage
  step: Partial<HttpRequestPlan>
}

export interface SummarizationJob {
  apiCallPlan: ApiCallPlan,
  contentId: string,
  chatId: string
  index: number
}

export const mockSequence = [
  {
    id: v4(),
    progressStage: ProgressStage.active,
    step: {
      name: 'Search for Artist “Skrillex”: Retrieve Skrillex’s artist ID by searching for his name using the Spotify API.',
      httpVerb: 'GET',
      url: 'https://api.spotify.com/v1/search',
      queryParams: {
        q: 'Skrillex',
        type: 'artist',
        limit: '1',
      },
      headers: {
        Authorization: 'Bearer {access_token}',
        'Content-Type': 'application/json',
      },
    },
  },
  {
    id: v4(),
    progressStage: ProgressStage.draft,
    step: {
      name: `Get Skrillex’s Top Tracks: Use the artist ID to fetch his top tracks from Spotify`,
      httpVerb: 'GET',
      url: 'https://api.spotify.com/v1/artists/{artist_id}/top-tracks',
      queryParams: {
        country: 'US',
      },
      headers: {
        Authorization: 'Bearer {access_token}',
        'Content-Type': 'application/json',
      },
    },
  },
  {
    id: v4(),
    progressStage: ProgressStage.draft,
    step: {
      name: `Get Your Spotify User ID: Obtain your user ID by accessing your Spotify profile information.`,
      httpVerb: 'GET',
      url: 'https://api.spotify.com/v1/me',
      headers: {
        Authorization: 'Bearer {access_token}',
        'Content-Type': 'application/json',
      },
    },
  },
  {
    id: v4(),
    progressStage: ProgressStage.draft,
    step: {
      name: `Create a New Playlist: Create a new playlist in your account to hold Skrillex’s top tracks.`,
      httpVerb: 'POST',
      url: 'https://api.spotify.com/v1/users/{user_id}/playlists',
      headers: {
        Authorization: 'Bearer {access_token}',
        'Content-Type': 'application/json',
      },
      body: {
        name: 'Skrillex Top Tracks',
        description: 'A playlist of top Skrillex tracks',
        public: true,
      },
    },
  },
  {
    id: v4(),
    progressStage: ProgressStage.draft,
    step: {
      name: `Add Tracks to the Playlist: Add the retrieved top tracks to your new playlist.`,
      httpVerb: 'POST',
      url: 'https://api.spotify.com/v1/playlists/{playlist_id}/tracks',
      headers: {
        Authorization: 'Bearer {access_token}',
        'Content-Type': 'application/json',
      },
      body: {
        uris: ['spotify:track:1', 'spotify:track:2', '...'], // Replace with actual track URIs
      },
    },
  },
] satisfies SequenceActivity[]
