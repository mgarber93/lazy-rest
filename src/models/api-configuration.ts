import {ClientOptions} from 'openai'


export interface ApiConfiguration {
  fileHandle: string;
  name: string;
  baseUrl: string;
  clientId: string;
  clientSecret: string;
}

export interface BedrockConfiguration {
  accessKey: string;
  secretAccessKey: string;
  accessToken?: string;
}

export interface ProviderConfiguration {
  openAi: ClientOptions | null;
  bedrock: BedrockConfiguration | null;
}
