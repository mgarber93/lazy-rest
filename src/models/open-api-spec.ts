export interface OpenApiSpec {
  openapi: "3.0.3"
  info: {
    description: string
    title: string
  }
  servers: { url: string } []
  tags: { name: string }[]
  paths: Record<string, Endpoint>
  components: {
    securitySchemes: any
    responses: any
    schemas: any
    parameters: any
  }
}

export interface Endpoint {
  get?: {
    tags: string[]
    description: string
  }
  post?: {
    tags: string[]
    description: string
  }
  put?: {
    tags: string[]
    description: string
  }
  delete?: {
    tags: string[]
    description: string
  }
}
