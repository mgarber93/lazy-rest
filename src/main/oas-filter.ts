import {OpenApiSpec} from '../models/open-api-spec';

function setEndpointDescription(object: Record<string, any>, path: string, key: string, value: any) {
  if (!(path in object)) {
    object[path] = {};
  }
  object[path][key] = value.replace('\n', ' ').trim();
}

export function oasToDescriptions(oasSpec: OpenApiSpec): object {
  const spec = {};
  for (const key in oasSpec.paths) {
    const endpoint = oasSpec.paths[key];
    
    if (endpoint.get?.description) {
      setEndpointDescription(spec, key, 'get', endpoint.get.description);
    }
    if (endpoint.put?.description) {
      setEndpointDescription(spec, key, 'put', endpoint.put.description);
    }
    if (endpoint.patch?.description) {
      setEndpointDescription(spec, key, 'patch', endpoint.patch.description);
    }
    if (endpoint.post?.description) {
      setEndpointDescription(spec, key, 'post', endpoint.post.description);
    }
    if (endpoint.delete?.description) {
      setEndpointDescription(spec, key, 'delete', endpoint.delete.description);
    }
  }
  return spec;
}