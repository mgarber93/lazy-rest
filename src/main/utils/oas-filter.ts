import {Endpoint, OpenApiSpec} from '../../models/open-api-spec'
import {HttpRequestPlan} from '../../models/conversation'

function setEndpointDescription(object: Record<string, any>, path: string, key: string, value: any) {
  if (!(path in object)) {
    object[path] = {}
  }
  object[path][key] = value.replace('\n', ' ').trim()
}

/**
 * gets the description of paths in the specification
 * It does not read the description of the oas overall, which is also available
 * @param oasSpec
 */
export function oasToDescriptions(oasSpec: OpenApiSpec): object {
  const spec = {}
  for (const key in oasSpec.paths) {
    const endpoint = oasSpec.paths[key]
    if (endpoint.get?.description) {
      setEndpointDescription(spec, key, 'get', endpoint.get.description)
    }
    if (endpoint.put?.description) {
      setEndpointDescription(spec, key, 'put', endpoint.put.description)
    }
    if (endpoint.patch?.description) {
      setEndpointDescription(spec, key, 'patch', endpoint.patch.description)
    }
    if (endpoint.post?.description) {
      setEndpointDescription(spec, key, 'post', endpoint.post.description)
    }
    if (endpoint.delete?.description) {
      setEndpointDescription(spec, key, 'delete', endpoint.delete.description)
    }
  }
  return spec
}

export function fuzzyMatch(a: string, b: string): boolean {
  const aSegments = a.split("/")
  const bSegments = b.split("/")
  
  if (aSegments.length !== bSegments.length) {
    return false
  }
  
  for (let i = 0; i < aSegments.length; i++) {
    const aSegment = aSegments[i]
    const bSegment = bSegments[i]
    // if a segment is surrounded by curly braces, acts like a wild card
    if (aSegment.startsWith("{") && aSegment.endsWith("}")) {
      continue
    }
    if (bSegment.startsWith("{") && bSegment.endsWith("}")) {
      continue
    }
    if (aSegments[i] !== bSegments[i]) {
      return false
    }
  }
  return true
}

export function treeShake(oasSpec: OpenApiSpec, plans: HttpRequestPlan[]) {
  const spec = {} as Record<string, any>
  for (const endpointPath in oasSpec.paths) {
    for (const plan of plans) {
      if (fuzzyMatch(endpointPath, plan.path)) {
        const matched = oasSpec.paths[endpointPath]
        const verb = plan.method.toLowerCase()
        spec[endpointPath] = {}
        spec[endpointPath][verb] = matched[verb as keyof Endpoint]
      }
    }
  }
  return spec
}