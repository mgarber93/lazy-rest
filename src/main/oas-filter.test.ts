import {oasToDescriptions} from './oas-filter';
import spec from '../../oas/spotify-oas.json';
import {OpenApiSpec} from '../models/open-api-spec';

test('oasToDescriptions', () => {
  const actual = oasToDescriptions(spec as OpenApiSpec);
  expect(actual).toMatchSnapshot();
});