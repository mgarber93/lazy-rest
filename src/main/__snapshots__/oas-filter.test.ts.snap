// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oasToDescriptions 1`] = `
{
  "/albums": {
    "get": "Get Spotify catalog information for multiple albums identified by their Spotify IDs.",
  },
  "/albums/{id}": {
    "get": "Get Spotify catalog information for a single album.",
  },
  "/albums/{id}/tracks": {
    "get": "Get Spotify catalog information about an album’s tracks. Optional parameters can be used to limit the number of tracks returned.",
  },
  "/artists": {
    "get": "Get Spotify catalog information for several artists based on their Spotify IDs.",
  },
  "/artists/{id}": {
    "get": "Get Spotify catalog information for a single artist identified by their unique Spotify ID.",
  },
  "/artists/{id}/albums": {
    "get": "Get Spotify catalog information about an artist's albums.",
  },
  "/artists/{id}/related-artists": {
    "get": "Get Spotify catalog information about artists similar to a given artist. Similarity is based on analysis of the Spotify community's listening history.",
  },
  "/artists/{id}/top-tracks": {
    "get": "Get Spotify catalog information about an artist's top tracks by country.",
  },
  "/audio-analysis/{id}": {
    "get": "Get a low-level audio analysis for a track in the Spotify catalog. The audio analysis describes the track’s structure and musical content, including rhythm, pitch, and timbre.",
  },
  "/audio-features": {
    "get": "Get audio features for multiple tracks based on their Spotify IDs.",
  },
  "/audio-features/{id}": {
    "get": "Get audio feature information for a single track identified by its unique Spotify ID.",
  },
  "/audiobooks": {
    "get": "Get Spotify catalog information for several audiobooks identified by their Spotify IDs. Audiobooks are only available within the US, UK, Canada, Ireland, New Zealand and Australia markets.",
  },
  "/audiobooks/{id}": {
    "get": "Get Spotify catalog information for a single audiobook. Audiobooks are only available within the US, UK, Canada, Ireland, New Zealand and Australia markets.",
  },
  "/audiobooks/{id}/chapters": {
    "get": "Get Spotify catalog information about an audiobook's chapters. Audiobooks are only available within the US, UK, Canada, Ireland, New Zealand and Australia markets.",
  },
  "/browse/categories": {
    "get": "Get a list of categories used to tag items in Spotify (on, for example, the Spotify player’s “Browse” tab).",
  },
  "/browse/categories/{category_id}": {
    "get": "Get a single category used to tag items in Spotify (on, for example, the Spotify player’s “Browse” tab).",
  },
  "/browse/categories/{category_id}/playlists": {
    "get": "Get a list of Spotify playlists tagged with a particular category.",
  },
  "/browse/featured-playlists": {
    "get": "Get a list of Spotify featured playlists (shown, for example, on a Spotify player's 'Browse' tab).",
  },
  "/browse/new-releases": {
    "get": "Get a list of new album releases featured in Spotify (shown, for example, on a Spotify player’s “Browse” tab).",
  },
  "/chapters": {
    "get": "Get Spotify catalog information for several audiobook chapters identified by their Spotify IDs. Chapters are only available within the US, UK, Canada, Ireland, New Zealand and Australia markets.",
  },
  "/chapters/{id}": {
    "get": "Get Spotify catalog information for a single audiobook chapter. Chapters are only available within the US, UK, Canada, Ireland, New Zealand and Australia markets.",
  },
  "/episodes": {
    "get": "Get Spotify catalog information for several episodes based on their Spotify IDs.",
  },
  "/episodes/{id}": {
    "get": "Get Spotify catalog information for a single episode identified by its unique Spotify ID.",
  },
  "/markets": {
    "get": "Get the list of markets where Spotify is available.",
  },
  "/me": {
    "get": "Get detailed profile information about the current user (including the current user's username).",
  },
  "/me/albums": {
    "delete": "Remove one or more albums from the current user's 'Your Music' library.",
    "get": "Get a list of the albums saved in the current Spotify user's 'Your Music' library.",
    "put": "Save one or more albums to the current user's 'Your Music' library.",
  },
  "/me/albums/contains": {
    "get": "Check if one or more albums is already saved in the current Spotify user's 'Your Music' library.",
  },
  "/me/audiobooks": {
    "delete": "Remove one or more audiobooks from the Spotify user's library.",
    "get": "Get a list of the audiobooks saved in the current Spotify user's 'Your Music' library.",
    "put": "Save one or more audiobooks to the current Spotify user's library.",
  },
  "/me/audiobooks/contains": {
    "get": "Check if one or more audiobooks are already saved in the current Spotify user's library.",
  },
  "/me/episodes": {
    "delete": "Remove one or more episodes from the current user's library.<br/> This API endpoint is in __beta__ and could change without warning. Please share any feedback that you have, or issues that you discover, in our [developer community forum](https://community.spotify.com/t5/Spotify-for-Developers/bd-p/Spotify_Developer).",
    "get": "Get a list of the episodes saved in the current Spotify user's library.<br/> This API endpoint is in __beta__ and could change without warning. Please share any feedback that you have, or issues that you discover, in our [developer community forum](https://community.spotify.com/t5/Spotify-for-Developers/bd-p/Spotify_Developer).",
    "put": "Save one or more episodes to the current user's library.<br/> This API endpoint is in __beta__ and could change without warning. Please share any feedback that you have, or issues that you discover, in our [developer community forum](https://community.spotify.com/t5/Spotify-for-Developers/bd-p/Spotify_Developer).",
  },
  "/me/episodes/contains": {
    "get": "Check if one or more episodes is already saved in the current Spotify user's 'Your Episodes' library.<br/> This API endpoint is in __beta__ and could change without warning. Please share any feedback that you have, or issues that you discover, in our [developer community forum](https://community.spotify.com/t5/Spotify-for-Developers/bd-p/Spotify_Developer)..",
  },
  "/me/following": {
    "delete": "Remove the current user as a follower of one or more artists or other Spotify users.",
    "get": "Get the current user's followed artists.",
    "put": "Add the current user as a follower of one or more artists or other Spotify users.",
  },
  "/me/following/contains": {
    "get": "Check to see if the current user is following one or more artists or other Spotify users.",
  },
  "/me/player": {
    "get": "Get information about the user’s current playback state, including track or episode, progress, and active device.",
    "put": "Transfer playback to a new device and optionally begin playback. This API only works for users who have Spotify Premium. The order of execution is not guaranteed when you use this API with other Player API endpoints.",
  },
  "/me/player/currently-playing": {
    "get": "Get the object currently being played on the user's Spotify account.",
  },
  "/me/player/devices": {
    "get": "Get information about a user’s available Spotify Connect devices. Some device models are not supported and will not be listed in the API response.",
  },
  "/me/player/next": {
    "post": "Skips to next track in the user’s queue. This API only works for users who have Spotify Premium. The order of execution is not guaranteed when you use this API with other Player API endpoints.",
  },
  "/me/player/pause": {
    "put": "Pause playback on the user's account. This API only works for users who have Spotify Premium. The order of execution is not guaranteed when you use this API with other Player API endpoints.",
  },
  "/me/player/play": {
    "put": "Start a new context or resume current playback on the user's active device. This API only works for users who have Spotify Premium. The order of execution is not guaranteed when you use this API with other Player API endpoints.",
  },
  "/me/player/previous": {
    "post": "Skips to previous track in the user’s queue. This API only works for users who have Spotify Premium. The order of execution is not guaranteed when you use this API with other Player API endpoints.",
  },
  "/me/player/queue": {
    "get": "Get the list of objects that make up the user's queue.",
    "post": "Add an item to the end of the user's current playback queue. This API only works for users who have Spotify Premium. The order of execution is not guaranteed when you use this API with other Player API endpoints.",
  },
  "/me/player/recently-played": {
    "get": "Get tracks from the current user's recently played tracks. _**Note**: Currently doesn't support podcast episodes._",
  },
  "/me/player/repeat": {
    "put": "Set the repeat mode for the user's playback. This API only works for users who have Spotify Premium. The order of execution is not guaranteed when you use this API with other Player API endpoints.",
  },
  "/me/player/seek": {
    "put": "Seeks to the given position in the user’s currently playing track. This API only works for users who have Spotify Premium. The order of execution is not guaranteed when you use this API with other Player API endpoints.",
  },
  "/me/player/shuffle": {
    "put": "Toggle shuffle on or off for user’s playback. This API only works for users who have Spotify Premium. The order of execution is not guaranteed when you use this API with other Player API endpoints.",
  },
  "/me/player/volume": {
    "put": "Set the volume for the user’s current playback device. This API only works for users who have Spotify Premium. The order of execution is not guaranteed when you use this API with other Player API endpoints.",
  },
  "/me/playlists": {
    "get": "Get a list of the playlists owned or followed by the current Spotify user.",
  },
  "/me/shows": {
    "delete": "Delete one or more shows from current Spotify user's library.",
    "get": "Get a list of shows saved in the current Spotify user's library. Optional parameters can be used to limit the number of shows returned.",
    "put": "Save one or more shows to current Spotify user's library.",
  },
  "/me/shows/contains": {
    "get": "Check if one or more shows is already saved in the current Spotify user's library.",
  },
  "/me/top/{type}": {
    "get": "Get the current user's top artists or tracks based on calculated affinity.",
  },
  "/me/tracks": {
    "delete": "Remove one or more tracks from the current user's 'Your Music' library.",
    "get": "Get a list of the songs saved in the current Spotify user's 'Your Music' library.",
    "put": "Save one or more tracks to the current user's 'Your Music' library.",
  },
  "/me/tracks/contains": {
    "get": "Check if one or more tracks is already saved in the current Spotify user's 'Your Music' library.",
  },
  "/playlists/{playlist_id}": {
    "get": "Get a playlist owned by a Spotify user.",
    "put": "Change a playlist's name and public/private state. (The user must, of course, own the playlist.)",
  },
  "/playlists/{playlist_id}/followers": {
    "delete": "Remove the current user as a follower of a playlist.",
    "put": "Add the current user as a follower of a playlist.",
  },
  "/playlists/{playlist_id}/followers/contains": {
    "get": "Check to see if one or more Spotify users are following a specified playlist.",
  },
  "/playlists/{playlist_id}/images": {
    "get": "Get the current image associated with a specific playlist.",
    "put": "Replace the image used to represent a specific playlist.",
  },
  "/playlists/{playlist_id}/tracks": {
    "delete": "Remove one or more items from a user's playlist.",
    "get": "Get full details of the items of a playlist owned by a Spotify user.",
    "post": "Add one or more items to a user's playlist.",
    "put": "Either reorder or replace items in a playlist depending on the request's parameters. To reorder items, include \`range_start\`, \`insert_before\`, \`range_length\` and \`snapshot_id\` in the request's body.
To replace items, include \`uris\` as either a query parameter or in the request's body.
Replacing items in a playlist will overwrite its existing items. This operation can be used for replacing or clearing items in a playlist.
<br/>
**Note**: Replace and reorder are mutually exclusive operations which share the same endpoint, but have different parameters.
These operations can't be applied together in a single request.",
  },
  "/recommendations": {
    "get": "Recommendations are generated based on the available information for a given seed entity and matched against similar artists and tracks. If there is sufficient information about the provided seeds, a list of tracks will be returned together with pool size details. 
For artists and tracks that are very new or obscure there might not be enough data to generate a list of tracks.",
  },
  "/recommendations/available-genre-seeds": {
    "get": "Retrieve a list of available genres seed parameter values for [recommendations](/documentation/web-api/reference/get-recommendations).",
  },
  "/search": {
    "get": "Get Spotify catalog information about albums, artists, playlists, tracks, shows, episodes or audiobooks that match a keyword string. Audiobooks are only available within the US, UK, Canada, Ireland, New Zealand and Australia markets.",
  },
  "/shows": {
    "get": "Get Spotify catalog information for several shows based on their Spotify IDs.",
  },
  "/shows/{id}": {
    "get": "Get Spotify catalog information for a single show identified by its unique Spotify ID.",
  },
  "/shows/{id}/episodes": {
    "get": "Get Spotify catalog information about an show’s episodes. Optional parameters can be used to limit the number of episodes returned.",
  },
  "/tracks": {
    "get": "Get Spotify catalog information for multiple tracks based on their Spotify IDs.",
  },
  "/tracks/{id}": {
    "get": "Get Spotify catalog information for a single track identified by its unique Spotify ID.",
  },
  "/users/{user_id}": {
    "get": "Get public profile information about a Spotify user.",
  },
  "/users/{user_id}/playlists": {
    "get": "Get a list of the playlists owned or followed by a Spotify user.",
    "post": "Create a playlist for a Spotify user. (The playlist will be empty until you [add tracks](/documentation/web-api/reference/add-tracks-to-playlist).)
Each user is generally limited to a maximum of 11000 playlists.",
  },
}
`;
